payload/reggae/rules/cmake.d

```diff
@@ -0,0 +1,320 @@
+        IncludePaths includes;
+        if ("includes" in compileGroup) {
+            includes = compileGroup["includes"].array
+                                    .map!(jobj => jobj["path"].str)
+                                    .array.IncludePaths;
+        }
+        return groupIndexToOptions[groupIndex] = CompileOptions(Flags(defines ~ " " ~ compileCommand),
+                                                                includes,
+                                                                cast(SourceLanguage) compileGroup["language"].str);
+    }
+    return getCompileOptionsImpl(compileGroup, compileGroupIndex);
+}
+
+private bool isCppHeader(in string file) {
+    import std.path : extension;
+    string ext = file.extension;
+    return ext == ".hpp" || ext == ".hxx" || ext == ".H";
+}
+
+private bool isCHeader(in string file) {
+    import std.path : extension;
+    return file.extension == ".h";
+}
+
+private bool isHeaderFile(in string file) {
+    return isCHeader(file) || isCppHeader(file);
+}
+
+private string getLinkerFlags(in imported!"std.json".JSONValue target) {
+    import std.array : join;
+    import std.algorithm : each;
+
+    string flags, libraries, libraryPath;
+
+    if ("link" !in target) {
+        return null;
+    }
+
+    auto link = target["link"];
+    link["commandFragments"].array.each!((jobj) {
+        switch(jobj["role"].str) {
+            case "flags":
+                flags ~= jobj["fragment"].str ~ " ";
+                break;
+            case "libraries":
+                libraries ~= jobj["fragment"].str ~ " ";
+                break;
+            case "libraryPath":
+                libraryPath ~= "-L" ~ jobj["fragment"].str ~ " ";
+                break;
+            default:
+                throw new Exception("Linker flag of type '" ~ jobj["role"].str ~ "' is not supported.");
+        }
+    });
+
+    return [flags, libraries, libraryPath].join(" ");
+}
+
+private enum TargetType : string {
+    Executable = "EXECUTABLE",
+    StaticLib = "STATIC_LIBRARY",
+    SharedLib = "SHARED_LIBRARY",
+}
+
+private bool isSupportedTarget(in string target) {
+    import std.traits : EnumMembers;
+    import std.algorithm : canFind;
+    return [EnumMembers!TargetType].canFind(target);
+}
+
+private imported!"reggae.build".Target toReggaeTarget(in imported!"std.json".JSONValue target,
+                                                      CMakeInfo cmakeInfo) {
+    import reggae.config: options;
+    import reggae.rules.common : objectFile, link, staticLibraryTarget;
+    import reggae.build : Target;
+    import reggae.types : SourceFile, ExeName, Flags;
+    import std.format : format;
+    import std.exception : enforce;
+    import std.path : buildNormalizedPath, isAbsolute, baseName;
+    import std.range : empty;
+    import std.stdio : writeln, stderr;
+
+    const sourceDir = cmakeInfo.codeModel["paths"]["source"].str;
+    const buildDir = cmakeInfo.codeModel["paths"]["source"].str;
```

Comment:
Is this right?

Reply:
```
paths
A JSON object containing members:

  source
  A string specifying the absolute path to the top-level source directory, represented with forward slashes.
  
  build
  A string specifying the absolute path to the top-level build directory, represented with forward slashes.
```

Reply:
I meant: do we want the build directory to be the same as the project directory?

Reply:
It is not the same.

https://github.com/atilaneves/reggae/pull/251/files#diff-17f07ad8161b4117a369be3e230df315d1410630362956eccda077bb3169517cR291

Reply:
But `sourceDir` and `buildDir` are the same?